File: September/22/validnumber.cpp
// LeetCode 65: Valid Number
// Explanation:
// 1. Use flags to track digit, decimal, and exponent positions.
// 2. Validate step by step similar to a finite state machine.
// Time Complexity: O(n)
// Space Complexity: O(1)

class Solution {
public:
    bool isNumber(string s) {
        int n = s.size();
        int i = 0;
        while (i < n && isspace(s[i])) i++; // trim leading spaces
        if (i < n && (s[i] == '+' || s[i] == '-')) i++; // sign

        bool num = false, dot = false, exp = false;

        for (; i < n; i++) {
            if (isdigit(s[i])) {
                num = true;
            } else if (s[i] == '.') {
                if (dot || exp) return false;
                dot = true;
            } else if (s[i] == 'e' || s[i] == 'E') {
                if (exp || !num) return false;
                exp = true;
                num = false;
                if (i+1 < n && (s[i+1] == '+' || s[i+1] == '-')) i++;
            } else if (isspace(s[i])) {
                while (i < n && isspace(s[i])) i++;
                return i == n && num;
            } else {
                return false;
            }
        }
        return num;
    }
};

File: September/22/validnumber.py
# LeetCode 65: Valid Number
# Explanation:
# 1. Use a finite state machine or flags to track digit, exponent, decimal, and sign validity.
# 2. Iterate through string and validate each character.
# Time Complexity: O(n)
# Space Complexity: O(1)

class Solution:
    def isNumber(self, s: str) -> bool:
        s = s.strip()
        num, dot, exp = False, False, False
        for i, ch in enumerate(s):
            if ch.isdigit():
                num = True
            elif ch in ['+', '-']:
                if i > 0 and s[i-1] not in ['e', 'E']:
                    return False
            elif ch == '.':
                if dot or exp:
                    return False
                dot = True
            elif ch in ['e', 'E']:
                if exp or not num:
                    return False
                exp, num = True, False
            else:
                return False
        return num

    def isNumberBuiltIn(self, s: str) -> bool:
        pattern = re.compile(r'^[+-]?((\d+(\.\d*)?)|(\.\d+))(e[+-]?\d+)?$', re.I)
        return bool(pattern.match(s.strip()))
# Example usage:
# sol = Solution()
# print(sol.isNumber("0"))      # True
# print(sol.isNumber("e"))      # False
# print(sol.isNumber("2e10"))   # True
